merge dholstius-Rcpp with master

dholstius opened the pull request.

dholstius committed:
make NAMESPACE export fmatch

dholstius committed:
add tests to master

dholstius committed:
add scratch-fmatch-timings (and increase SAMPLE_SIZES)

dholstius committed:
move scratch/ to root dir

dholstius committed:
add real-world test data

dholstius committed:
promote scratch/ content to vignette

dholstius committed:
improve comments

dholstius committed:
improve str_wcmp()

- check for end of s2, not just s1
- make wildcard a parameter

dholstius committed:
improve as_c_str()

use strdup() instead of rolling our own

dholstius committed:
remove superfluous export

dholstius committed:
tidy whitespace

dholstius committed:
construct r with NA_INTEGER directly

dholstius committed:
tidy comment strings

dholstius committed:
take CharacterVector instead of std::vector< std::string >

dholstius committed:
add packages to Suggests

dholstius committed:
version bump

dholstius committed:
move vignettes/ to root dir

dholstius committed:
Revert "take CharacterVector instead of std::vector< std::string >"

This reverts commit de5b74a67fc2b0fced0b3d568da0a9b5faa47a2e.

dholstius committed:
reversion of de5b74a

dholstius committed:
add VignetteBuilder: knitr

dholstius committed:
fix bug in vignette

dholstius committed:
add library(ggtools) for log10 scale prettiness

dholstius committed:
echo vignette output to console

dholstius committed:
ignore .R files in vignettes/

dholstius committed:
add empty final line to DESCRIPTION

dholstius committed:
replace x.size() with x_size

dholstius committed:
index with square brackets

dholstius committed:
index with square brackets

dholstius committed:
Merge branch 'dholstius-Rcpp' of https://github.com/BAAQMD/fmatch into dholstius-Rcpp

Commit:
Fix crash bug in vignette.

Commit:
Suppress irrelvant package messages

dholstius committed:
Merge pull request #2 from BAAQMD/alamstein-Rcpp

dholstius commented:
@arilamstein, will you give this a once-over and then merge? Thank you!

arilamstein commented on NAMESPACE:
```diff
@@ -1,3 +1,3 @@
-# Generated by roxygen2: do not edit by hand
```
@dholstius This is a minor thing, but could cause issues later on. Note that this comment ("# Generated by roygen2: do not edit by hand") is in the old version of this file, not the current version. This makes it impossible to use `roxygen2::roxygenize()` in the future.

At the command line, when I type:
```
> library(roxygen2)
> roxygenize()
```
I now get this output:
```
ℹ Loading fmatch
Warning: The existing 'NAMESPACE' file was not generated by roxygen2, and will not be overwritten.
Warning message:
roxygen2 requires Encoding: UTF-8 
```

Is this intentional? My intuition is that the current `NAMESPACE` file was, indeed created by `roxygen2`. Perhaps during our last meeting you modified it while showing me something about how Rcpp works in the NAMESPACE file and accidentally checked it in?

If this is a mistake, I recommend deleting the current NAMESPACE file, recreating it with `roxygen2::roxygenize()` and then commiting the new file in.

arilamstein reviewed (commented):
@dholstius I'm not sure if you got a notification when I created this comment yesterday (or was it 2 days ago?) but I think that there might be an issue with the NAMESPACE file. It looks like github's PR interface has changed since I last used it, and perhaps you only get notifications when I click "submit review". 🤷‍♂️  

Also, I submitted a PR to this branch yesterday which has an implementation of the algorithm that does not require the creation of extra data structures. I will leave it to you whether to go ahead with this PR or to review/implement my code first.

However, irrespective of my PR, I would appreciate clarity around this NAMESPACE / roxygen2 issue before approving the PR. If it is an accident, we should fix it. If it is intentional / necessary due to Rcpp, then it's something I should be aware of.

dholstius committed:
add Encoding: UTF-8

dholstius committed:
make Rcpp and roxygen play nice together

following https://r-pkgs.org/src.html#cpp , I ran `usethis::use_rcpp()` and then followed the directions (copy-pasted below):

```
> usethis::use_rcpp()
✓ Setting active project to '/Users/dholstius/GitHub/BAAQMD/fmatch'
✓ Adding '*.o', '*.so', '*.dll' to 'src/.gitignore'
● Copy and paste the following lines into 'R/fmatch-package.R':
  ## usethis namespace: start
  #' @useDynLib fmatch, .registration = TRUE
  ## usethis namespace: end
  NULL
  [Copied to clipboard]
✓ Adding 'Rcpp' to LinkingTo field in DESCRIPTION
✓ Adding 'Rcpp' to Imports field in DESCRIPTION
● Copy and paste the following lines into 'R/fmatch-package.R':
  ## usethis namespace: start
  #' @importFrom Rcpp sourceCpp
  ## usethis namespace: end
  NULL
  [Copied to clipboard]
```

dholstius committed:
use precomputed size and also use square brackets

Merge branch 'dholstius-Rcpp' of https://github.com/BAAQMD/fmatch into dholstius-Rcpp

# Conflicts:
#	DESCRIPTION

dholstius committed:
export fmatch()

dholstius committed:
setup vignette quietly

dholstius commented:
@arilamstein, thanks for the review! Fixed the roxygen/Rcpp issue (with the `NAMESPACE`) in the most recent commits. Anything else catch your eye?

dholstius committed:
version bump

dholstius committed:
Merge branch 'dholstius-Rcpp' of https://github.com/BAAQMD/fmatch into dholstius-Rcpp

dholstius commented:
Oops, one last edit. ↑

arilamstein commented:
On a whim I tried to do a CHECK and it generated an error in the vignette. I then tried to build the vignette by hand (i.e. clicking `Knit` in `fmatch_performance.Rmd` and it generates this error:
```
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──
✓ ggplot2 3.3.3     ✓ purrr   0.3.4
✓ tibble  3.1.0     ✓ dplyr   1.0.5
✓ tidyr   1.1.3     ✓ stringr 1.4.0
✓ readr   1.4.0     ✓ forcats 0.5.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()
Loading required package: funtools
Loading required package: rlang

Attaching package: 'rlang'

The following objects are masked from 'package:purrr':

    %@%, as_function, flatten, flatten_chr, flatten_dbl, flatten_int,
    flatten_lgl, flatten_raw, invoke, list_along, modify, prepend,
    splice


Attaching package: 'funtools'

The following object is masked from 'package:rlang':

    is_true

Loading required package: strtools
Loading required package: scales

Attaching package: 'scales'

The following object is masked from 'package:purrr':

    discard

The following object is masked from 'package:readr':

    col_factor

Loading required package: tbltools
Loading required package: codec
Loading required package: packtools
Loading required package: vartools
Loading required package: difftools
Loading required package: daff
Loading required package: diffobj
Loading required package: lubridate

Attaching package: 'lubridate'

The following objects are masked from 'package:base':

    date, intersect, setdiff, union

Loading required package: readxl
Loading required package: rpivotTable

Error: package or namespace load failed for 'tbltools':
 object 'fmatch' is not exported by 'namespace:fmatch'
Quitting from lines 13-18 (fmatch_performance.Rmd) 
Error: package 'tbltools' could not be loaded
Execution halted
```
I'm not sure exactly what is going on here (I just started looking into it).

My first impression is that it's probably OK if some of our internal tools fail R CMD CHECK for obscure reasons. But we just invested some time making the vignette work, and it it did work until recently, so it's probably worth trying to fix this. I will continue to investigate.

arilamstein commented:
Quick update: I just verified that this is a "real" error. I cannot `library(tbltools)` with fmatch v0.3. But the command does work when fmatch v0.1 is loaded. That means that we probably cannot use this version.

This is my first project using Rcpp. But my guess is that there is something funky that goes on with Rcpp and roxygen2. That's where I'm starting to look.

arilamstein commented:
I mean, this is the NAMESPACE for v1.0: https://github.com/BAAQMD/fmatch/blob/master/NAMESPACE
```
# Generated by roxygen2: do not edit by hand

export(fmatch)
```

And this is the (current, broken one):
```
# Generated by roxygen2: do not edit by hand

export(fmatch)
importFrom(Rcpp,sourceCpp)
useDynLib(fmatch, .registration = TRUE)
```

I don't know what the final 2 lines do. But the export is the same and is something I'm accustomed to seeing in R package NAMESPACE files.

arilamstein commented:
Interestingly, when I nuke your NAMESPACE file and recreate it myself it generates this file:
```
# Generated by roxygen2: do not edit by hand

importFrom(Rcpp,sourceCpp)
useDynLib(fmatch, .registration = TRUE)
```

I.e. no explicit "export".

And when I type:
```
> library(fmatch)
> fmatch
Error: object 'fmatch' not found
> ?fmatch # this works - it brings up the help
> fmatch("a","b")
Error in fmatch("a", "b") : could not find function "fmatch"
```
According to the package viewer, fmatch 0.3 is loaded (not 0.1).

arilamstein commented on R/RcppExports.R:
```diff
@@ -14,6 +14,6 @@
 #' @export
 #' 
 fmatch <- function(x, y) {
-    .Call('_fmatch_fmatch', PACKAGE = 'fmatch', x, y)
```
@dholstius I did a google search for "roxygen and Rcpp" and wound up here: https://gallery.rcpp.org/articles/documenting-rcpp-packages/. It says

> When using Rcpp in package development, the R functions in RcppExports.R are generated automatically from the source .cpp files. Rcpp also faithfully transcribes comment blocks from .cpp to .R. We can take advantage of this fact to add documentation directly to the .cpp files and have it appear in the package reference manual.

Based on that I looked for differences in this file in the two versions. I noticed that the `PACKAGE` parameter is missing. I wonder if that could be related to our bug.

arilamstein commented:
I just removed all my changes and attempted to do a clean R CMD CHECK and it failed with this error (which, I believe, is different than the errors I reported yesterday):
```
> * checking package dependencies ... ERROR
 > VignetteBuilder package not declared: ‘knitr’
```
You can fix this by adding `knitr` to the suggests line. But then there are other errors.

I think we should adopt an explicit policy regarding whether internal packages need to pass R CMD CHECK or not. My first thought is that there are some pedantic errors this process complains about which are not important. Nonetheless, passing these tests will give us a degree of comfort / defensibility that our packages are "playing nice" with others. And so it is probably "worth it" to at least try and pass all these errors.

arilamstein commented:
Also, let me know if you want me to submit changes like this (adding `knitr` to DESCRIPTION) myself. I can't tell if right now you want me to be in just the role of reviewer or also in the role of developer. And so I'm taking the most conservative approach (and just providing feedback).

arilamstein commented:
A follow up with the Rcpp / export stuff I originally reported yesterday. I nuked all local changes this morning and started from scratch to try to get a fresh perspective on the errors.

What I noticed is that when I sync to the head of this branch, and click either "Install and Restart" or "Check" the first thing it seems to do is call `Rcpp::compileAttributes()`. This, in turn, modifies `R/RcppExports.R`. It add these lines to the file:
```
#' @export
#'
```
To the extent that you wanted a code review, my question is: "Should this PR already contain these two lines?"

dholstius commented:
> "Should this PR already contain these two lines?"

I see that the `@export` line(s) are in fact in `R/RcppExports.R` (below) "as is" — are you sure you're working with the head of the `dholstius-Rcpp` branch? If you are, I'm confused!

https://github.com/BAAQMD/fmatch/blob/a18150d17a59f218a19050b6996f299a27059b09/R/RcppExports.R#L13-L19

dholstius committed:
move strings.Rds and patterns.Rds to extdata/

- fixes problem with building vignette

dholstius committed:
add packages to Suggests

dholstius committed:
make check() happy with package description

dholstius committed:
version bump

dholstius commented:
For me, the `dholstius-Rcpp` branch now passes `check()` cleanly, with only a minor note:

```
> checking installed package size ... NOTE
    installed size is  5.3Mb
    sub-directories of 1Mb or more:
      extdata   4.2Mb

0 errors ✓ | 0 warnings ✓ | 1 note x
```

Please have another go at your convenience!

dholstius commented:
> I think we should adopt an explicit policy regarding whether internal packages need to pass R CMD CHECK or not. My first thought is that there are some pedantic errors this process complains about which are not important. Nonetheless, passing these tests will give us a degree of comfort / defensibility that our packages are "playing nice" with others. And so it is probably "worth it" to at least try and pass all these errors.

I like the idea of a policy that internal packages should pass `check()` cleanly (or with reasonable cleanliness, as above). At a minimum, there should be no errors produced by `check()`.

This will entail a goodly amount of tidying-up work, as historically I've been more oriented toward moving fast. (That was an explicit decision by Phil and I: to let some technical debt accumulate.) So, for example, a lot of parameters are mismatched with the documentation, or undocumented. And sometimes we will have to deploy with known bugs that don't affect the current project heap.

But I like it!

arilamstein commented:
> For me, the `dholstius-Rcpp` branch now passes `check()` cleanly, with only a minor note:
> 
> ```
> > checking installed package size ... NOTE
>     installed size is  5.3Mb
>     sub-directories of 1Mb or more:
>       extdata   4.2Mb
> 
> 0 errors ✓ | 0 warnings ✓ | 1 note x
> ```
> 
> Please have another go at your convenience!

I just sync'd to the head of `dholstius-Rcpp` and unfortunately when I run R CMD CHECK I get 2 warnings and 1 note. In addition to the NOTE you describe above, I also get:
```
    WARNING
   ‘qpdf’ is needed for checks on size reduction of PDFs

   Error(s) in re-building vignettes:
   Error: package or namespace load failed for 'tbltools' in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
    there is no package called 'data.table'
   Quitting from lines 13-18 (fmatch_performance.Rmd) 
   Error: processing vignette 'fmatch_performance.Rmd' failed with diagnostics:
   package 'tbltools' could not be loaded
   --- failed re-building ‘fmatch_performance.Rmd’
   
   SUMMARY: processing the following file failed:
     ‘fmatch_performance.Rmd’
   
   Error: Vignette re-building failed.
   Execution halted
```

I am using the Server. I wonder if you are possibly using your home machine to test? The additional warnings I get are themselves not the end of the world, but I do consider it disconcerting that our outputs differ at all.

Also, regarding the differences in `R/RcppExports.R`. I can confirm that when I sync, my file is exactly like the file in github ([link](https://github.com/BAAQMD/fmatch/blob/dholstius-Rcpp/R/RcppExports.R)). That is, line 14 has the `@export` call. But whenever I click "Build" or "Check" the `@export` line disappears. I don't have a good explanation for that. 

I think it might be best if we schedule a time to pair program through this. What do you think? Perhaps you will see something on my setup that explains this. The largest concern, I think, is that if you and I have differing outputs, then we can't be sure what users of the package will experience.

arilamstein commented:
> > I think we should adopt an explicit policy regarding whether internal packages need to pass R CMD CHECK or not. My first thought is that there are some pedantic errors this process complains about which are not important. Nonetheless, passing these tests will give us a degree of comfort / defensibility that our packages are "playing nice" with others. And so it is probably "worth it" to at least try and pass all these errors.
> 
> I like the idea of a policy that internal packages should pass `check()` cleanly (or with reasonable cleanliness, as above). At a minimum, there should be no errors produced by `check()`.
> 
> This will entail a goodly amount of tidying-up work, as historically I've been more oriented toward moving fast. (That was an explicit decision by Phil and I: to let some technical debt accumulate.) So, for example, a lot of parameters are mismatched with the documentation, or undocumented. And sometimes we will have to deploy with known bugs that don't affect the current project heap.
> 
> But I like it!

This is something that I can help you with. I'm not sure how it fits with the scope of my current contract, but I have a lot of experience writing production-quality code, and I'd be happy to help.

dholstius commented:
Yes, let's schedule a time to pair-program through it. I'm on leave next week. How about Apr 6 in the afternoon?

You're right! I was on my laptop. I should have been running `check()` on the server. 🤦‍♂️ Will do.

dholstius commented:
Can you paste the output of `devtools::session_info()`? So we can isolate any library-related differences. I'm away from my desk, but will also do so soon.

dholstius commented:
Which exact commit are you working with? Just to be sure.

dholstius commented:
> I'm not sure how it fits with the scope of my current contract

I think there are provisions in there that would cover helping out with "hardening" of in-house R package infrastructure; at least, I had that in mind when helping Song and Phil draft the SOW. Happy to discuss!

arilamstein commented:
> Can you paste the output of `devtools::session_info()`? So we can isolate any library-related differences. I'm away from my desk, but will also do so soon.

Good idea. Here is my output:
```
> devtools::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.3 (2020-02-29)
 os       Ubuntu 18.04.5 LTS          
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language (EN)                        
 collate  C.UTF-8                     
 ctype    C.UTF-8                     
 tz       America/Los_Angeles         
 date     2021-03-26                  

─ Packages ─────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date       lib source        
 assertthat    0.2.1   2019-03-21 [2] CRAN (R 3.6.3)
 cachem        1.0.4   2021-02-13 [2] CRAN (R 3.6.3)
 callr         3.5.1   2020-10-13 [2] CRAN (R 3.6.3)
 cli           2.3.1   2021-02-23 [2] CRAN (R 3.6.3)
 crayon        1.4.1   2021-02-08 [2] CRAN (R 3.6.3)
 desc          1.2.0   2018-05-01 [2] CRAN (R 3.6.3)
 devtools      2.3.2   2020-09-18 [2] CRAN (R 3.6.3)
 ellipsis      0.3.1   2020-05-15 [2] CRAN (R 3.6.3)
 fastmap       1.1.0   2021-01-25 [2] CRAN (R 3.6.3)
 fs            1.5.0   2020-07-31 [2] CRAN (R 3.6.3)
 glue          1.4.2   2020-08-27 [2] CRAN (R 3.6.3)
 lifecycle     1.0.0   2021-02-15 [2] CRAN (R 3.6.3)
 magrittr      2.0.1   2020-11-17 [2] CRAN (R 3.6.3)
 memoise       2.0.0   2021-01-26 [2] CRAN (R 3.6.3)
 pkgbuild      1.2.0   2020-12-15 [2] CRAN (R 3.6.3)
 pkgload       1.2.0   2021-02-23 [2] CRAN (R 3.6.3)
 prettyunits   1.1.1   2020-01-24 [2] CRAN (R 3.6.3)
 processx      3.4.5   2020-11-30 [2] CRAN (R 3.6.3)
 ps            1.6.0   2021-02-28 [2] CRAN (R 3.6.3)
 purrr         0.3.4   2020-04-17 [2] CRAN (R 3.6.3)
 R6            2.5.0   2020-10-28 [2] CRAN (R 3.6.3)
 remotes       2.2.0   2020-07-21 [2] CRAN (R 3.6.3)
 rlang         0.4.10  2020-12-30 [2] CRAN (R 3.6.3)
 rprojroot     2.0.2   2020-11-15 [2] CRAN (R 3.6.3)
 sessioninfo   1.1.1   2018-11-05 [2] CRAN (R 3.6.3)
 testthat      3.0.2   2021-02-14 [2] CRAN (R 3.6.3)
 usethis       2.0.1   2021-02-10 [2] CRAN (R 3.6.3)
 withr         2.4.1   2021-01-26 [2] CRAN (R 3.6.3)

[1] /aim/users/ari/R/x86_64-pc-linux-gnu-library/3.6
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library
```

It occured to me during this process that updating the version of `R` and calling `update.packages()` might solve this, or at least lead to different results. That's my default when I have problems building my own packages. But the server environment makes that tricky to implement.

I will not that `Rcpp` and `roxygen2` are not listed above. But their versions are 1.0.6 and 7.1.1, respectively. When I click the "Install and Restart" button in RStudio, the first line is:
```
==> Rcpp::compileAttributes()

* Updated R/RcppExports.R
```
so that package is executed as part of this process, even though it is not "loaded".

arilamstein commented:
> Which exact commit are you working with? Just to be sure.
```
ari@baaqmd-aim-rstudio:~/github/fmatch$ git log -1 --oneline
0e52e1a (HEAD -> dholstius-Rcpp, origin/dholstius-Rcpp) version bump
```

Also, it appears to be the latest version:
```
ari@baaqmd-aim-rstudio:~/github/fmatch$ git pull
Already up to date.
```

arilamstein commented:
The afternoon of April 6 works for me!

On Fri, Mar 26, 2021 at 4:03 PM David Holstius ***@***.***>
wrote:

> Yes, let's schedule a time to pair-program through it. I'm on leave next
> week. How about Apr 6 in the afternoon?
>
> You're right! I was on my laptop. I should have been checking on the
> server. Will do.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BAAQMD/fmatch/pull/3#issuecomment-808578522>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMMYEJ2IPCDPY4MHISKSTTFUHFLANCNFSM4ZUFK22A>
> .
>


-- 
Ari Lamstein
R Training and Consulting
Website: AriLamstein.com
Cell: 510-499-7212

arilamstein commented:
> > I'm not sure how it fits with the scope of my current contract
> 
> I think there are provisions in there that would cover helping out with "hardening" of in-house R package infrastructure; at least, I had that in mind when helping Song and Phil draft the SOW. Happy to discuss!

This is good to know. Let's plan to talk about it a bit more on April 6!

dholstius committed:
add roxygen export in fmatch.cpp docstring

dholstius commented:
OK, now I've also (on the server):
- reinstalled the `data.table` package
- installed the `qpdf`, `downloader`, and `googleVis` packages

Some of those are most likely in the `Depends:` section of an in-house R package when they could probably be in the `Imports:`. Housekeeping to do!

dholstius committed:
version bump

dholstius commented:
Finally, I installed `qpdf` on the system itself, via:

```
sudo apt-get install qpdf
```

... and the relevant warning from `check()` ("‘qpdf’ is needed for checks on size reduction of PDFs") went away.

So, now `check()` works cleanly for me on the server, with the exception of the little "note":

```
> checking installed package size ... NOTE
   installed size is  6.0Mb
   sub-directories of 1Mb or more:
     extdata   4.2Mb
```

... which I'm OK with. And, `devtools::install(build_vignettes = TRUE)` also succeeds, with no errors or warnings.

dholstius commented:
For the record, here's my `session_info()` on the server:

```
> devtools::session_info()
─ Session info ────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.3 (2020-02-29)
 os       Ubuntu 18.04.5 LTS          
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language (EN)                        
 collate  C.UTF-8                     
 ctype    C.UTF-8                     
 tz       America/Los_Angeles         
 date     2021-03-26                  

─ Packages ────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date       lib source        
 assertthat    0.2.1   2019-03-21 [2] CRAN (R 3.6.3)
 cachem        1.0.4   2021-02-13 [2] CRAN (R 3.6.3)
 callr         3.5.1   2020-10-13 [2] CRAN (R 3.6.3)
 cli           2.3.1   2021-02-23 [2] CRAN (R 3.6.3)
 crayon        1.4.1   2021-02-08 [2] CRAN (R 3.6.3)
 curl          4.3     2019-12-02 [2] CRAN (R 3.6.3)
 desc          1.2.0   2018-05-01 [2] CRAN (R 3.6.3)
 devtools    * 2.3.2   2020-09-18 [2] CRAN (R 3.6.3)
 digest        0.6.27  2020-10-24 [2] CRAN (R 3.6.3)
 ellipsis      0.3.1   2020-05-15 [2] CRAN (R 3.6.3)
 fastmap       1.1.0   2021-01-25 [2] CRAN (R 3.6.3)
 fmatch      * 0.3.2   2021-03-27 [1] Github (BAAQMD/fmatch@7d48337)
 fs            1.5.0   2020-07-31 [2] CRAN (R 3.6.3)
 glue          1.4.2   2020-08-27 [2] CRAN (R 3.6.3)
 knitr         1.31    2021-01-27 [2] CRAN (R 3.6.3)
 lifecycle     1.0.0   2021-02-15 [2] CRAN (R 3.6.3)
 magrittr      2.0.1   2020-11-17 [2] CRAN (R 3.6.3)
 memoise       2.0.0   2021-01-26 [2] CRAN (R 3.6.3)
 pkgbuild      1.2.0   2020-12-15 [2] CRAN (R 3.6.3)
 pkgload       1.2.0   2021-02-23 [2] CRAN (R 3.6.3)
 prettyunits   1.1.1   2020-01-24 [2] CRAN (R 3.6.3)
 processx      3.4.5   2020-11-30 [2] CRAN (R 3.6.3)
 ps            1.6.0   2021-02-28 [2] CRAN (R 3.6.3)
 purrr         0.3.4   2020-04-17 [2] CRAN (R 3.6.3)
 R6            2.5.0   2020-10-28 [2] CRAN (R 3.6.3)
 rcmdcheck     1.3.3   2019-05-07 [2] CRAN (R 3.6.3)
 Rcpp          1.0.6   2021-01-15 [2] CRAN (R 3.6.3)
 remotes       2.2.0   2020-07-21 [2] CRAN (R 3.6.3)
 rlang         0.4.10  2020-12-30 [2] CRAN (R 3.6.3)
 roxygen2      7.1.1   2020-06-27 [2] CRAN (R 3.6.3)
 rprojroot     2.0.2   2020-11-15 [2] CRAN (R 3.6.3)
 rstudioapi    0.13    2020-11-12 [2] CRAN (R 3.6.3)
 sessioninfo   1.1.1   2018-11-05 [2] CRAN (R 3.6.3)
 stringi       1.5.3   2020-09-09 [2] CRAN (R 3.6.3)
 stringr       1.4.0   2019-02-10 [2] CRAN (R 3.6.3)
 testthat    * 3.0.2   2021-02-14 [2] CRAN (R 3.6.3)
 usethis     * 2.0.1   2021-02-10 [2] CRAN (R 3.6.3)
 withr         2.4.1   2021-01-26 [2] CRAN (R 3.6.3)
 xfun          0.22    2021-03-11 [2] CRAN (R 3.6.3)
 xml2          1.3.2   2020-04-23 [2] CRAN (R 3.6.3)
 xopen         1.0.0   2018-09-17 [2] CRAN (R 3.6.3)

[1] /aim/users/dholstius/R/x86_64-pc-linux-gnu-library/3.6
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library
```

arilamstein commented:
This is looking better. But after syncing I still see different results from R CMD CHECK than you. I see:
```
  Error: package or namespace load failed for 'tbltools' in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
   there is no package called 'data.table'
  Quitting from lines 13-18 (fmatch_performance.Rmd) 
  Error: processing vignette 'fmatch_performance.Rmd' failed with diagnostics:
  package 'tbltools' could not be loaded
  --- failed re-building ‘fmatch_performance.Rmd’
```
This is odd because I see that the `data.table` is installed. On a whim, I decided to directly add `library(data.table)` as the first line of the code block in the vignette that loads all the packages. I get this error:
```
there is no package called 'data.table'
  --- failed re-building ‘fmatch_performance.Rmd’
```
However, if I type `library(data.table)` on the console of course it works. 

It is probably worth us spending at least an hour or two (a) trying to figure out why I get this error but you don't and (b) how to resolve it for me. I did a quick google search and the first few links didn't help.

arilamstein commented:
I think I have a better understanding of the issue. I think that the vignette is probably sort of limited in the packages it can see. In particular, I think it is limited to the packages listed in the package's `DESCRIPTION`.  Here is that list:
```
Suggests: 
    testthat,
    inventory,
    tidyverse,
    knitr,
    here,
    microbenchmark,
    ggplot2,
    tidyr,
    ggtool
Remotes:
  github::BAAQMD/ggtools
```
I note that `tbltools` is that not there. As an experiment, I added `data.table` to the `SUGGESTS` list and the existing error about `data.table` went away, but it was replaced by a new package it complained was missing: `downloader`.

I tried to add `github::BAAQMD/tbltools` to the `Remotes` section but it didn't make the error go away. I assume that that is because the repo is private.

If I'm right about this (that the error is due to a required library being private, which `knitr` can't access), then we can probably safely ignore the error. But I am still confused as to why you don't seem to have this error.

dholstius commented:
Have you restarted your R session since I installed `downloader` and a couple of other packages? (I left a comment in the PR about doing that ... maybe I mistakenly installed them in my user library, instead of the site library? Yuan could help you install them in the site library, if so.)

arilamstein commented:
> Have you restarted your R session since I installed `downloader` and a couple of other packages? (I left a comment in the PR about doing that ... maybe I mistakenly installed them in my user library, instead of the site library? Yuan could help you install them in the site library, if so.)

Yes, I restarted R. The vignette works perfectly for me when I hit the "knit" button. But it fails to create as part of R CMD Check. I think the failure is due to the vignette loading a package (`tbltools`) that the package "claims" it does not need (i.e. it is not mentioned in the DESCRIPTION in Suggests, Imports or Depends).

arilamstein commented:
Per my recent meeting with @dholstius : We now both get the same minor error related to building the vignette with R CMD CHECK. Note that the vignette works fine when built manually.

I will accept the PR as-is and create a new issue for that problem. The cost-benefit seems to strongly favor getting this new version out.

dholstius commented:
:shipit:
